### ISSUE TITLE
Problem with validation of identical fields in the schema

### ISSUE THREAD

**Issue by gorinovsoft** [colinhacks/zod#3752]
Currently, there is an issue in the schema where, if identical fields are present in the user's schemas, Zod incorrectly validates the second testProperty field. Is there a way to fix this? At the moment, the only solution I've found is to change the field name (testProperty) to a unique one.

[https://codesandbox.io/p/sandbox/zodschemas-forked-76tlcr?file=%2Fsrc%2FApp.tsx
![Снимок экрана 2024-09-12 115322](https://github.com/user-attachments/assets/cf91d4f4-25a7-4871-a04f-c3e078e56620)
](url)

`import { useCallback, VFC } from "react";
import { SubmitHandler, useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { z } from "zod";

const stringSchema = z.string().min(1, { message: "fieldIsRequired" });

const firstUser = z.object({
  firstName: stringSchema,
  lastName: stringSchema,
  testProperty: stringSchema,
});

const secondUser = z.object({
  firstName: stringSchema,
  lastName: stringSchema,
  testProperty: stringSchema.or(z.string().length(0)),
  testPropertySecond: stringSchema.or(z.string().length(0)),
});

const UserSchema = z.union([firstUser, secondUser]);

const usersSchema = z.object({
  users: UserSchema.array(),
});

type UsersForm = z.infer<typeof usersSchema>;

export const App: VFC = () => {
  const { register, handleSubmit, formState, control } = useForm<UsersForm>({
    resolver: zodResolver(usersSchema),
    defaultValues: {
      users: [
        {
          firstName: "",
          lastName: "",
          testProperty: "",
        },
        {
          firstName: "",
          lastName: "",
          testProperty: "",
          testPropertySecond: "",
        },
      ],
    },
  });

  const errors = formState.errors;

  console.log(errors, formState, control);

  const onSubmit: SubmitHandler<UsersForm> = useCallback(async (value) => {},
  []);

  return (
    <form
      onSubmit={handleSubmit(onSubmit)}
      style={{
        display: "flex",
        flexDirection: "column",
        width: 512,
        margin: "0 auto",
      }}
    >
      <label>
        <span>First name first user</span>
        <input {...register("users.0.firstName")} />
        <span>{errors?.users?.[0]?.firstName?.message}</span>
      </label>
      <label>
        <span>Last name first user</span>
        <input {...register("users.0.lastName")} />
        <span>{errors?.users?.[0]?.lastName?.message}</span>
      </label>
      <label>
        <span>Test property first user</span>
        <input {...register("users.0.testProperty")} />
        <span>{errors?.users?.[0]?.testProperty?.message}</span>
      </label>

      <label style={{ marginTop: "40px" }}>
        <span>First name second user</span>
        <input {...register("users.1.firstName")} />
        <span>{errors?.users?.[1]?.firstName?.message}</span>
      </label>
      <label>
        <span>Last name second user</span>
        <input {...register("users.1.lastName")} />
        <span>{errors?.users?.[1]?.lastName?.message}</span>
      </label>
      <label>
        <span>Test property second user</span>
        <input {...register("users.1.testProperty")} />
        <span>{errors?.users?.[1]?.testProperty?.message}</span>
      </label>
      <label>
        <span>Second test property second user</span>
        <input {...register("users.1.testPropertySecond")} />
        <span>{errors?.users?.[1]?.testPropertySecond?.message}</span>
      </label>
      <button type="submit">Submit</button>
    </form>
  );
};
`

**Comment by Amirika18** [comment]
@sunnylost it could be a work solution, but if we change the order of zod schemas in array and then decide to add property with the same key but with different validation schemas, validation will raise error in `testPropertySecond` property. Look at this example below: 
![image](https://github.com/user-attachments/assets/a92f42c2-0689-4a24-9921-cebefeb3617f)

**Comment by sunnylost** [comment]
So the issue is that you want to use the `firstUser` schema to parse the first user and the `secondUser` schema for the second, but since both users have the same structure, there's no way to determine which schema to apply. In my opinion, each user should have a property to distinguish them, allowing you to use a discriminated union for validation.

**Comment by Amirika18** [comment]
> So the issue is that you want to use the `firstUser` schema to parse the first user and the `secondUser` schema for the second, but since both users have the same structure, there's no way to determine which schema to apply. In my opinion, each user should have a property to distinguish them, allowing you to use a discriminated union for validation.

Yes, you are right! We want to apply the schema to object depends on some key (for example, id from enum). Unfortunately, zod union do not support the determination of validation schema according to key property and its value. We have `z.discriminatedUnion` which could help if it could work with transformed zod schemas.
